### ISSUE TITLE
Allow filtering by createdAt/updatedAt

### ISSUE THREAD

**Issue by chrisDeFouRire** [strapi/strapi#3595]
It's currently not possible to search with the createdAt and updatedAt fields, because these fields not included here: https://github.com/strapi/strapi/blob/master/packages/strapi-utils/lib/buildQuery.js#L8

**Steps to reproduce the behavior**
`curl 'http://localhost:1337/whatever?createdAt_gt=2019-07-10T12:11:51.402Z'` will fail with the following message: "Your filters contain a field 'createdAt' that doesn't appear on your model definition nor it's relations"

**Expected behavior**
Searching by createdAt and updatedAt should work, it's a common feature.
It would be even better if we could also use HTTP fields like `If-Modified-Since`...

**Code snippets**

```const isAttribute = (model, field) => _.has(model.allAttributes, field) || model.primaryKey === field || field == "createdAt" || field == "updatedAt";``` 
would probably fix the issue.

**System**
- Node.js version: official docker image
- NPM version:
- Strapi version: 3.0.0-alpha-26.2
- Database: mongodb
- Operating system: linux

**Additional context**

I need to use strapi as a "mostly static" CMS, and will keep data in a local cache (VERY high query rate): updatedAt and createdAt would help a lot for cache invalidation.

**Comment by derrickmehaffy** [comment]
@chrisDeFouRire your code snippet would normally work however there is an option to change the automated timestamp field names. I'm fairly certain they are including in searches (at least in my testing with my project on MySQL) 

With Bookshelf/Knex based databases the field names are `created_at` and `updated_at`:

![image](https://user-images.githubusercontent.com/8593673/60975065-1fb76f00-a2e0-11e9-803a-65c747fde4f6.png)

(the above screenshot is based on beta.10)

**Comment by derrickmehaffy** [comment]
The easiest way to confirm this (as that code you referenced in buildQuery, I believe uses this internal route) is using `strapi console` and checking what Strapi lists as attributes of the model. It is possible that the mongo part is not correctly pulling `allAttributes` and is instead just using `attributes` which won't have the automated timestamp fields included.

![image](https://user-images.githubusercontent.com/8593673/60975313-a1a79800-a2e0-11e9-8d28-3ee0f3cae5c3.png)

**Comment by chrisDeFouRire** [comment]
Thank you @derrickmehaffy, I tried the strapi console:

```
strapi > strapi.models.vgood.allAttributes
{ costs: { type: 'json' },
  description: { required: true, type: 'string' },
  disabled: { type: 'boolean' },
  name: { required: true, type: 'string' },
  propertySet: { type: 'json' },
  shortCode: { required: true, type: 'string', unique: true },
  tags: { type: 'json' },
  type: { type: 'string' } }
```

no `createdAt` or `updatedAt`

**Comment by derrickmehaffy** [comment]
@chrisDeFouRire can you copy and paste your `api/vgood/models/Vgood.settings.json` file here

Wondering if you don't have the timestamps enabled on that model :thinking:

**Comment by chrisDeFouRire** [comment]
@derrickmehaffy here it is:
```
{
  "connection": "default",
  "collectionName": "vgood",
  "info": {
    "name": "vgood",
    "description": ""
  },
  "options": {
    "timestamps": true
  },
  "attributes": {
    "costs": {
      "type": "json"
    },
    "description": {
      "required": true,
      "type": "string"
    },
    "disabled": {
      "type": "boolean"
    },
    "name": {
      "required": true,
      "type": "string"
    },
    "propertySet": {
      "type": "json"
    },
    "shortCode": {
      "required": true,
      "type": "string",
      "unique": true
    },
    "tags": {
      "type": "json"
    },
    "type": {
      "type": "string"
    }
  }
}```

**Comment by derrickmehaffy** [comment]
they are enabled for sure `"timestamps": true`

I would say this is a confirmed bug then, but will need to confirm with some more testing. @lauriejim

**Comment by chrisDeFouRire** [comment]
Note: I didn't write a single line of code, it's 100% Content Type Builder.

From the look of https://github.com/strapi/strapi/blob/master/packages/strapi-hook-mongoose/lib/index.js#L345 and the previous lines, I think the timestamp fields are never added to `allAttributes`

**Comment by chrisDeFouRire** [comment]
@derrickmehaffy Here's how the timestamp attributes ARE added in the bookshelf hook:

https://github.com/strapi/strapi/blob/master/packages/strapi-hook-bookshelf/lib/buildDatabaseSchema.js#L259

```
  // Add created_at and updated_at field if timestamp option is true
  if (loadedModel.hasTimestamps) {
    definition.attributes[
      _.isString(loadedModel.hasTimestamps[0])
        ? loadedModel.hasTimestamps[0]
        : 'created_at'
    ] = {
      type: 'timestamp',
    };
    definition.attributes[
      _.isString(loadedModel.hasTimestamps[1])
        ? loadedModel.hasTimestamps[1]
        : 'updated_at'
    ] = {
      type: 'timestampUpdate',
    };
  }

  // Save all attributes (with timestamps)
  model.allAttributes = _.clone(definition.attributes);
```
(this last line is also present in the mongoose hook)

**Comment by derrickmehaffy** [comment]
@chrisDeFouRire can you try changing your model to use this:
```
    "timestamps": [
      "createdAt",
      "updatedAt"
    ],
```

Instead of the `timestamps: true` bit

**Comment by chrisDeFouRire** [comment]
I tried... didn't work, same result (as setting to true is the same as setting this array).
I think the problem is the missing addition to definition.attributes to add the timestamps with their type

**Comment by alexandrebodin** [comment]
@chrisDeFouRire I see you are in alpha and reference code in beta. Have you tried the beta version ?

**Comment by derrickmehaffy** [comment]
Ah snap nice catch @alexandrebodin I didn't even look at that. @chrisDeFouRire I don't think filtering by those timestamps was supported in the alpha

**Comment by chrisDeFouRire** [comment]
@alexandrebodin I'm using the docker image which is alpha... however even the code in master doesn't contain the following, which DOES fix it for me:


```
                    if (_.isArray(_.get(definition, 'options.timestamps'))) {
                      definition.attributes[_.get(definition, 'options.timestamps[0]')] = {
                        type: 'timestamp',
                      };
                      definition.attributes[_.get(definition, 'options.timestamps[1')] = {
                        type: 'timestampUpdate',
                      };
                    }                  

                    // Save all attributes (with timestamps)
                    target[model].allAttributes = _.clone(definition.attributes);
```

This code should go here: https://github.com/strapi/strapi/blob/master/packages/strapi-hook-mongoose/lib/index.js#L344

And now the model looks just fine:

```
strapi > strapi.models.vgood.allAttributes
{ costs: { type: 'json' },
  description: { required: true, type: 'string' },
  disabled: { type: 'boolean' },
  name: { required: true, type: 'string' },
  propertySet: { type: 'json' },
  shortCode: { required: true, type: 'string', unique: true },
  tags: { type: 'json' },
  type: { type: 'string' },
  createdAt: { type: 'timestamp' },
  updatedAt: { type: 'timestampUpdate' } }
```

And most important, my query works fine too with this modification (hot fixed in node_modules, dirty AF but it shows the fix is the right one):

`curl 'http://server:1337/vgoods?createdAt_gt=2019-07-10T12:11:51.402Z'`

I can prepare a pull request if you need...

**Comment by derrickmehaffy** [comment]
please do a pull request :)

**Comment by derrickmehaffy** [comment]
@chrisDeFouRire can you try testing filtering on mongo with a fresh beta project?

**Comment by chrisDeFouRire** [comment]
> @chrisDeFouRire can you try testing filtering on mongo with a fresh beta project?

I have no idea how to build/test... I've not even used the CLI, only a docker image and the webapp.

Here's what I've done:
```
yarn build
yarn run test:generate-app
yarn run test:start-app

yarn test:unit
yarn test:front

```

also tried
```
cd testApp
./node_modules/.bin/strapi console
```

I did generate a test app with mongodb of course...
but for some reason I can't navigate to /admin, it goes to index.html no matter what.

So the console shows `testApp > [2019-07-11T07:09:47.097Z] debug GET index.html (27 ms)` when I navigate to `http://myserver.in.a.vm:1337/admin`.

So I can't test, because there's no admin I can use to create a model, populate it, to finally test.

The getStarted example doesn't use mongodb... it's not going to help
`[2019-07-11T07:38:45.046Z] warn (hook:mongoose) wasn't loaded due to missing key `enabled` in the configuration`

I was able to fix a bug, but not to run a simple test...
I can't say strapi makes it easy to onboard developers...

**Comment by derrickmehaffy** [comment]
For the beta and testing with the mono-repo (much of the documentation for this isn't quite complete yet)

You don't even need to install the global package if you use npx or yarn:

1. Create a normal beta project: `yarn create strapi-app test`
2. Clone the monorepo somewhere locally: `git clone git@github.com:strapi/strapi.git`
3. Navigate into the monorepo and run `yarn setup` then `node scripts/link.js`
4. The last command will dump some stuff to the console you'll need to copy that and paste it in your terminal while inside your project directory:

```
cd path/to/your/project && yarn link create-strapi-app strapi strapi-admin strapi-generate strapi-generate-api strapi-generate-controller strapi-generate-model strapi-generate-new strapi-generate-plugin strapi-generate-policy strapi-generate-service strapi-helper-plugin strapi-hook-bookshelf strapi-hook-ejs strapi-hook-knex strapi-hook-mongoose strapi-hook-redis strapi-middleware-views strapi-plugin-content-manager strapi-plugin-content-type-builder strapi-plugin-documentation strapi-plugin-email strapi-plugin-graphql strapi-plugin-settings-manager strapi-plugin-upload strapi-plugin-users-permissions strapi-provider-email-amazon-ses strapi-provider-email-mailgun strapi-provider-email-sendgrid strapi-provider-email-sendmail strapi-provider-upload-aws-s3 strapi-provider-upload-cloudinary strapi-provider-upload-local strapi-provider-upload-rackspace strapi-utils
```

You should get something like this: 
![image](https://user-images.githubusercontent.com/8593673/61033847-1af2c980-a379-11e9-8ac7-e82419fa7e17.png)

From there your packages are now using the monorepo ones instead of NPM ones, you can confirm this by looking at at your node_modules folder:
![image](https://user-images.githubusercontent.com/8593673/61034083-74f38f00-a379-11e9-87ff-9d71dc29b26e.png)

thus any changes you make in the monorepo will reflect in this project.

After you have linked them you run `yarn build` and `yarn start` / `yarn develop`

**Comment by derrickmehaffy** [comment]
(In that first screenshot I was in the wrong directory, I should have ran the yarn link inside the path to my project which was `~/Development/Strapi/test3`

**Comment by derrickmehaffy** [comment]
Yarn linked packages are usually stored in `~/.config/yarn/link` which contains symlinks to where ever you have them located at: 

![image](https://user-images.githubusercontent.com/8593673/61034444-209cdf00-a37a-11e9-8cae-2dba4fd98f1e.png)

**Comment by derrickmehaffy** [comment]
@alexandrebodin wasn't this fixed in beta.14?

**Comment by alexandrebodin** [comment]
nope this will be fixed in an upcoming release

**Comment by danielharven** [comment]
was this ever fixed am using 3.5.1 and can't sort by createdAt field
